go/vault/api/action.go

```diff
@@ -0,0 +1,360 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"slices"
+
+	"github.com/oasisprotocol/oasis-core/go/common"
+	"github.com/oasisprotocol/oasis-core/go/common/cbor"
+	"github.com/oasisprotocol/oasis-core/go/common/prettyprint"
+	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
+	staking "github.com/oasisprotocol/oasis-core/go/staking/api"
+)
+
+var _ prettyprint.PrettyPrinter = (*Action)(nil)
+
+// PendingAction is a pending action.
+type PendingAction struct {
+	// Nonce is the action nonce.
+	Nonce uint64 `json:"nonce"`
+	// AuthorizedBy contains the addresses that have authorized the action.
+	AuthorizedBy []staking.Address `json:"authorized_by"`
+	// Action is the pending action itself.
+	Action Action `json:"action"`
+}
+
+// ContainsAuthorizationFrom returns true iff the given address is among the action authorizers.
+func (pa *PendingAction) ContainsAuthorizationFrom(addr staking.Address) bool {
+	return slices.Contains(pa.AuthorizedBy, addr)
+}
+
+// Action is a vault action.
+type Action struct {
+	// Suspend is the suspend action.
+	Suspend *ActionSuspend `json:"suspend,omitempty"`
+	// Resume is the resume action.
+	Resume *ActionResume `json:"resume,omitempty"`
+	// ExecuteMessage is the execute message action.
+	ExecuteMessage *ActionExecuteMessage `json:"execute_msg,omitempty"`
+	// UpdateWithdrawPolicy is the withdraw policy update action.
+	UpdateWithdrawPolicy *ActionUpdateWithdrawPolicy `json:"update_withdraw_policy,omitempty"`
+	// UpdateAuthority is the authority update action.
+	UpdateAuthority *ActionUpdateAuthority `json:"update_authority,omitempty"`
+}
+
+// Validate validates the given action.
+func (a *Action) Validate(params *ConsensusParameters) error {
+	// Ensure there is at most one action set.
+	if !common.ExactlyOneTrue(
+		a.Suspend != nil,
+		a.Resume != nil,
+		a.ExecuteMessage != nil,
+		a.UpdateWithdrawPolicy != nil,
+		a.UpdateAuthority != nil,
+	) {
+		return fmt.Errorf("exactly one action must be set")
+	}
+
+	// Validate actions that need validation.
+	var err error
+	switch {
+	case a.ExecuteMessage != nil:
+		err = a.ExecuteMessage.Validate()
+	case a.UpdateWithdrawPolicy != nil:
+		err = a.UpdateWithdrawPolicy.Validate()
+	case a.UpdateAuthority != nil:
+		err = a.UpdateAuthority.Validate(params)
+	}
+	return err
+}
+
+// Equal returns true iff one action is equal to another.
+func (a *Action) Equal(other *Action) bool {
+	return bytes.Equal(cbor.Marshal(a), cbor.Marshal(other))
```

Comment:
Off-topic: It would be nice to have a method `cbor.Equal(a, b)` which could return as soon as encodings differ.